From bdb4e4b907a1649ab9f1017ec3c4c872616e8a16 Mon Sep 17 00:00:00 2001
From: Evgeny Voropaev <evorop@gmail.com>
Date: Wed, 21 Oct 2026 20:29:12 +0800
Subject: [PATCH] Exclude system attributes from recursive xid-format upgrade
 checks

Filter attnum > 0 in composite type propagation and final stored-column reporting. System xmin/xmax are handled by page conversion and must not reject every source table or contaminate safe row types.
---
 src/bin/pg_upgrade/check.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/src/bin/pg_upgrade/check.c b/src/bin/pg_upgrade/check.c
index c3b1516e9c0..8aaab74da06 100644
--- a/src/bin/pg_upgrade/check.c
+++ b/src/bin/pg_upgrade/check.c
@@ -375,6 +375,7 @@ data_type_check_query(int checknum)
 					"			WHERE t.typtype = 'c' AND "
 					"				  t.oid = c.reltype AND "
 					"				  c.oid = a.attrelid AND "
+					"				  a.attnum > 0 AND "
 					"				  NOT a.attisdropped AND "
 					"				  a.atttypid = x.oid "
 					"			UNION ALL "
@@ -389,6 +390,7 @@ data_type_check_query(int checknum)
 					"		pg_catalog.pg_namespace n, "
 					"		pg_catalog.pg_attribute a "
 					"WHERE	c.oid = a.attrelid AND "
+					"		a.attnum > 0 AND "
 					"		NOT a.attisdropped AND "
 					"		a.atttypid IN (SELECT oid FROM oids) AND "
 					"		c.relkind IN ("
-- 
2.55.0

